### PR TITLE
[5.6] brpoplpush for redis queues

### PR DESCRIPTION
**Pull request by halaei**
Fixing #22939
### Changes:
1. The order of jobs in Redis queue is reversed.
2. A new Redis list is added for each queue, prefiex ':front'.
3. brpoplpush command is used for blocking pop.
4. `LuaScripts::migrateExpiredJobs()` Migrates up to 100 jobs from reserved and delayed queues. It is not a good to eval lua scripts that are possible to take super-linear time. I think a PR may be sent to previous versions of Laravel with this regard.
### Tests:
1. I added a test for migrating more than 100 jobs, since the ordering was a bit confusing to me.
2. A test is added for blocking pop. It requires `pcntl_fork()`.

### CODE REVIEW

**Comment by sisve** [comment]
Could you explain why we would need another queue (the one with `:front`)? We already have a `$queue` and a `$queue:reserved`, why would we need a new one? And what happens to the `:reserved` queue with this change?

I guess I'm confused by the fact that there's a `$queue` and a `$queue:front` and we're popping jobs from both. Why have two keys for this?

**Comment by halaei** [comment]
queue:reserved is a sorted set that stores dequeued jobs after laravel-specific required modification. There is no redis command to do a blocking  pop from a list then do appropriate laravel specific modification and add it to a sorted set and there is no way to implement such a thing with a single lua script either. So basically a new queue:front list is added along queue:reserved sorted set to have a reliable laravel redis queue with blocking pop support.
I don't know if we can have a better implementation unless we write some [redis module in C](https://redis.io/topics/modules-intro).

**Comment by taylorotwell** [comment]
Do you know if it would be better to just maintain entirely different queues for blocking vs non-blocking and keep them in different formats? That way it isn't really a change for those using non-blocking? Just throwing out ideas.

**Comment by halaei** [comment]
I don't know how clean can we implement that.
To throw out some more ideas, I was also considering if someone can write us a redis module in C to make blocking pop reliable without adding a list or reverting the direction. That way we might even be able to do blocking pop from delayed/scheduled queues as I asked in [this feature request](https://github.com/antirez/redis/issues/2801) some years ago.
The problem is that if I want to do this it will probably take me more than a month since I know absolutely nothing about the details of redis code. I wish I could ask some experts like @antirez to kindly help us if the feature is highly demanded.

**Comment by mfn** [comment]
> write us a redis module in C

I feel this goes completely off-track here.

**Comment by halaei** [comment]
I looked at the module API. It sounds like blocking command part of the API is in experimental stage. It also seems that in order to have such module not only `RedisQueue::blockingPop()` but also all the other functions of `RedisQueue` class should be rewritten in C. So maybe @mfn is correct.

**Comment by sisve** [comment]
I see no obvious problems with this implementation, with the caveat that I have no experience from using BRPOPLPUSH myself.

I understand that we're reversing the queues, and it seems logical to me. I like that this PR changes order of both the blocking and the non-blocking popping list. It sounds reasonable that they process things in the same order. The change in order would be something that only affects people during the upgrade, and not when they keep switching between the two modes of popping (if I read the code correctly).

Regarding performance in the blocking workflow; does it matter at the moment? There are issues in the existing solution, and BRPOPLPUSH solves it. That is more important than if BRPOPLPUSH is slightly slower.

As I started saying; I've never used BRPOPLPUSH myself, so I may be missing something. It would be nice if we could get some redis-users with more experience in this area to look at the PR to verify that everything looks good.

I think it's more important to get a working "safe" solution merged and optimize it later on, than trying to find the perfect solution now.

File: tests/Queue/RedisQueueIntegrationTest.php [hunk 1]
```diff
@@ -335,10 +378,11 @@ public function testSize($driver)
 
     /**
      * @param string $driver
+     * @param null $blockFor
```
Comment (GrahamCampbell) [inline tests/Queue/RedisQueueIntegrationTest.php]: incorrect type doc

File: tests/Queue/RedisQueueIntegrationTest.php [hunk 2]
```diff
@@ -65,6 +65,49 @@ public function testExpiredJobsArePopped($driver)
...
+        if ($pid = pcntl_fork() > 0) {
+            $this->setUpRedis();
+            $this->setQueue($driver, 10);
+            $this->assertEquals(12, unserialize(json_decode($this->queue->pop()->getRawBody())->data->command)->i);
+        } elseif ($pid == 0) {
+            $this->setUpRedis();
+            $this->setQueue('predis');
+            sleep(1);
+            $this->queue->push(new RedisQueueIntegrationTestJob(12));
+            die;
```
Comment (GrahamCampbell) [inline tests/Queue/RedisQueueIntegrationTest.php]: ?

File: src/Illuminate/Queue/RedisQueue.php [hunk 3]
```diff
@@ -209,41 +209,41 @@ public function migrateExpiredJobs($from, $to)
...
+            return $this->nonBlockingPop($queue);
+        }
+
+        return [null, null];
+    }
+
+    /**
+     * Retrieve the next job without blocking.
+     *
+     * @param $queue
```
Comment (GrahamCampbell) [inline src/Illuminate/Queue/RedisQueue.php]: invalid syntax

**Review by GrahamCampbell — CHANGES_REQUESTED** [review]

File: tests/Queue/RedisQueueIntegrationTest.php (hunk 2 above)
Comment (halaei) [reply tests/Queue/RedisQueueIntegrationTest.php]: In order to test blocking pop, there should be a process (child) to push into queue while the worker process (parent) is waiting. After push the child process is done so it dies.

Comment (mfn) [reply tests/Queue/RedisQueueIntegrationTest.php]: If it doesn't matter, please use `exit("explaining why");`, think this is helpful for "the next developer". Alternatively, a comment in the source is welcome too. IMHO 😄

**Comment by taylorotwell** [comment]
I'm having a bit of trouble understanding the "front" concept too. What does it contain?

**Comment by halaei** [comment]
@taylorotwell The queue is devided into 2 parts, namely back and front. In Redis they are named "$queue" and "$queue:front", respectively.
When a process needs to push, it has to push to the back of the queue. When a worker pops into queue, it first pops from the front. If no item is there it pops from the back.
If both lists are empty and pops return nothing, the worker can wait for jobs via brpoplpush: it does a blocking pop from the back, waiting for a job to be pushed. Immediately after a job is pushed in the back, Redis moves the job to the front queue and wait the worker up. The worker now finds out there is a job inside the queue and safely removes it from there and reserve it into a zset via Lua script.

**Comment by taylorotwell** [comment]
For people who are already running queues on Redis on 5.6, what will their application do if they upgraded to this code?

**Comment by halaei** [comment]
If they upgrade to this code, jobs that were pushed before upgrade will be served in reverse order. So if the ordering is important to them, they need some after upgrade code to run once and fix the order for them, or they have to empty the queues before upgrade.

I sent this PR before release of 5.6. Will you merge it here or it should be resubmit to 5.7?

**Comment by mfn** [comment]
Also: ordering is "only" an issue for queues having only 1 worker.

**Comment by taylorotwell** [comment]
I'm just going to hold off on this for now.

**Comment by halaei** [comment]
@taylorotwell Please let me know if you need it in 5.7 or later. Thanks.

**Pull request closed without merging**